File: src/core/date_utils.cpp
// SPDX-FileCopyrightText: 2025 Sergio Martins
// SPDX-License-Identifier: MIT

#include "date_utils.h"
#include "Clock.h"

#include <array>
#include <chrono>
#include <cstdio>
#include <ctime>
#include <sstream>
#include <format>

namespace pointless::core::DateUtils {

namespace {
// Helper to convert time_point to std::tm
std::tm to_tm(std::chrono::system_clock::time_point tp)
{
    std::time_t t = std::chrono::system_clock::to_time_t(tp);
    std::tm tm {};
    ( void )localtime_r(&t, &tm);
    return tm;
}
}

std::string dateStr(std::chrono::system_clock::time_point date)
{
    std::tm tm = to_tm(date);
    std::ostringstream oss;
    oss << tm.tm_mday << "-" << (tm.tm_mon + 1) << "-" << (tm.tm_year + 1900);
    return oss.str();
}
std::string_view weekdayName(std::chrono::system_clock::time_point date)
{
    static const std::array<const char *, 7> names = { "Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday", "Sunday" };
    std::tm tm = to_tm(date);

    const int wday = tm.tm_wday == 0 ? 6 : tm.tm_wday - 1; // tm_wday: 0=Sunday
    return names.at(wday);
}

bool isWorkDay(std::chrono::system_clock::time_point date)
{
    std::tm tm = to_tm(date);
    return tm.tm_wday >= 1 && tm.tm_wday <= 5;
}

std::chrono::system_clock::time_point trimTime(std::chrono::system_clock::time_point date)
{
    std::tm tm = to_tm(date);
    tm.tm_hour = 0;
    tm.tm_min = 0;
    tm.tm_sec = 0;
    tm.tm_isdst = -1; // Let mktime determine DST
    std::time_t t = mktime(&tm);
    return std::chrono::system_clock::from_time_t(t);
}

std::chrono::system_clock::time_point today()
{
    return trimTime(Clock::now());
}

bool isToday(std::chrono::system_clock::time_point date)
{
    return trimTime(date) == today();
}

bool isTomorrow(std::chrono::system_clock::time_point date)
{
    auto yest = date - std::chrono::hours(24);
    return isToday(yest);
}

bool isYesterday(std::chrono::system_clock::time_point date)
{
    auto tom = date + std::chrono::hours(24);
    return isToday(tom);
}

std::chrono::system_clock::time_point thisWeeksMonday(std::chrono::system_clock::time_point date)
{
    std::tm tm = to_tm(date);
    int daysBack = (tm.tm_wday == 0 ? 6 : tm.tm_wday - 1);
    tm.tm_mday -= daysBack;
    tm.tm_hour = 0;
    tm.tm_min = 0;
    tm.tm_sec = 0;
    tm.tm_isdst = -1;
    std::time_t t = mktime(&tm);
    return std::chrono::system_clock::from_time_t(t);
}

bool isThisWeek(std::chrono::system_clock::time_point date)
{
    auto monday = thisWeeksMonday(Clock::now());
    return date >= monday && date < monday + std::chrono::hours(24 * 7);
}

bool isNext7Days(std::chrono::system_clock::time_point date)
{
    auto t = today();
    return (isToday(date) || isAfterToday(date)) && date < t + std::chrono::hours(24 * 7);
}

bool isAfterToday(std::chrono::system_clock::time_point date)
{
    return trimTime(date) > today();
}

bool iMidnight(std::chrono::system_clock::time_point date)
{
    std::tm tm = to_tm(date);
    return tm.tm_hour == 0 && tm.tm_min == 0 && tm.tm_sec == 0;
}

std::chrono::system_clock::time_point nextMonday(std::chrono::system_clock::time_point date)
{
    std::tm tm = to_tm(date);
    int daysBack = (tm.tm_wday == 0 ? 6 : tm.tm_wday - 1);
    tm.tm_mday += (7 - daysBack);
    tm.tm_hour = 0;
    tm.tm_min = 0;
    tm.tm_sec = 0;
    tm.tm_isdst = -1;
    std::time_t t = mktime(&tm);
    return std::chrono::system_clock::from_time_t(t);
}

std::string prettyDate(std::chrono::system_clock::time_point date, bool includeTime)
{
    if (isToday(date)) {
        return includeTime ? std::format("today {}", weekdayName(date)) : "today";
    }
    if (isTomorrow(date)) {
        return "tomorrow";
    }
    if (isYesterday(date)) {
        return "yesterday";
    }
    if (isThisWeek(date)) {
        return std::format("this {}", weekdayName(date));
    }
    if (isNext7Days(date)) {
        return std::format("next {}", weekdayName(date));
    }

    std::string result = dateStr(date);
    if (includeTime) {
        std::tm tm = to_tm(date);
        result += " ";
        result += std::format("{:02d}:{:02d}", tm.tm_hour, tm.tm_min);
    }
    return result;
}

} // namespace pointless::core::DateUtils

File: pointless_core/src/date_utils.rs
// SPDX-FileCopyrightText: 2025 Sergio Martins
// SPDX-License-Identifier: MIT

use chrono::{DateTime, Datelike, Duration, Local, NaiveTime, TimeZone, Timelike, Utc, Weekday};

use crate::clock;

pub fn trim_time(date: DateTime<Utc>) -> DateTime<Utc> {
    let local_date = date.with_timezone(&Local).date_naive();
    let midnight = NaiveTime::from_hms_opt(0, 0, 0).unwrap();
    let local_dt = local_date.and_time(midnight);
    Local
        .from_local_datetime(&local_dt)
        .single()
        .unwrap()
        .with_timezone(&Utc)
}

pub fn today() -> DateTime<Utc> {
    trim_time(clock::now())
}

pub fn is_today(date: DateTime<Utc>) -> bool {
    trim_time(date) == today()
}

pub fn is_tomorrow(date: DateTime<Utc>) -> bool {
    is_today(date - Duration::hours(24))
}

pub fn is_yesterday(date: DateTime<Utc>) -> bool {
    is_today(date + Duration::hours(24))
}

pub fn is_work_day(date: DateTime<Utc>) -> bool {
    let local = date.with_timezone(&Local);
    matches!(
        local.weekday(),
        Weekday::Mon | Weekday::Tue | Weekday::Wed | Weekday::Thu | Weekday::Fri
    )
}

pub fn is_midnight(date: DateTime<Utc>) -> bool {
    let local = date.with_timezone(&Local);
    local.hour() == 0 && local.minute() == 0 && local.second() == 0
}

pub fn this_weeks_monday(date: DateTime<Utc>) -> DateTime<Utc> {
    let local = date.with_timezone(&Local);
    let days_since_monday = local.weekday().num_days_from_monday();
    let monday = local.date_naive() - Duration::days(days_since_monday as i64);
    let midnight = NaiveTime::from_hms_opt(0, 0, 0).unwrap();
    Local
        .from_local_datetime(&monday.and_time(midnight))
        .single()
        .unwrap()
        .with_timezone(&Utc)
}

pub fn is_this_week(date: DateTime<Utc>) -> bool {
    let monday = this_weeks_monday(clock::now());
    date >= monday && date < monday + Duration::hours(24 * 7)
}

pub fn is_next_7_days(date: DateTime<Utc>) -> bool {
    let t = today();
    (is_today(date) || is_after_today(date)) && date < t + Duration::hours(24 * 7)
}

pub fn is_after_today(date: DateTime<Utc>) -> bool {
    trim_time(date) > today()
}

pub fn next_monday(date: DateTime<Utc>) -> DateTime<Utc> {
    let local = date.with_timezone(&Local);
    let days_since_monday = local.weekday().num_days_from_monday();
    let next_mon = local.date_naive() + Duration::days((7 - days_since_monday) as i64);
    let midnight = NaiveTime::from_hms_opt(0, 0, 0).unwrap();
    Local
        .from_local_datetime(&next_mon.and_time(midnight))
        .single()
        .unwrap()
        .with_timezone(&Utc)
}

pub fn weekday_name(date: DateTime<Utc>) -> &'static str {
    let local = date.with_timezone(&Local);
    match local.weekday() {
        Weekday::Mon => "Monday",
        Weekday::Tue => "Tuesday",
        Weekday::Wed => "Wednesday",
        Weekday::Thu => "Thursday",
        Weekday::Fri => "Friday",
        Weekday::Sat => "Saturday",
        Weekday::Sun => "Sunday",
    }
}

pub fn date_str(date: DateTime<Utc>) -> String {
    let local = date.with_timezone(&Local);
    format!("{}-{}-{}", local.day(), local.month(), local.year())
}

pub fn pretty_date(date: DateTime<Utc>, include_time: bool) -> String {
    if is_today(date) {
        return if include_time {
            format!("today {}", weekday_name(date))
        } else {
            "today".to_string()
        };
    }
    if is_tomorrow(date) {
        return "tomorrow".to_string();
    }
    if is_yesterday(date) {
        return "yesterday".to_string();
    }
    if is_this_week(date) {
        return format!("this {}", weekday_name(date));
    }
    if is_next_7_days(date) {
        return format!("next {}", weekday_name(date));
    }

    let mut result = date_str(date);
    if include_time {
        let local = date.with_timezone(&Local);
        result += &format!(" {:02}:{:02}", local.hour(), local.minute());
    }
    result
}

#[cfg(test)]
mod tests {
    use super::*;
    use chrono::TimeZone;

    fn make_utc(year: i32, month: u32, day: u32, hour: u32, min: u32, sec: u32) -> DateTime<Utc> {
        Utc.with_ymd_and_hms(year, month, day, hour, min, sec)
            .unwrap()
    }

    #[test]
    fn test_trim_time() {
        let dt = make_utc(2025, 6, 15, 14, 30, 45);
        let trimmed = trim_time(dt);
        let local = trimmed.with_timezone(&Local);
        assert_eq!(local.hour(), 0);
        assert_eq!(local.minute(), 0);
        assert_eq!(local.second(), 0);
    }

    #[test]
    fn test_is_today() {
        clock::set_test_now(make_utc(2025, 6, 15, 12, 0, 0));
        assert!(is_today(make_utc(2025, 6, 15, 8, 0, 0)));
        assert!(!is_today(make_utc(2025, 6, 14, 8, 0, 0)));
        clock::reset();
    }

    #[test]
    fn test_is_tomorrow() {
        clock::set_test_now(make_utc(2025, 6, 15, 12, 0, 0));
        assert!(is_tomorrow(make_utc(2025, 6, 16, 8, 0, 0)));
        assert!(!is_tomorrow(make_utc(2025, 6, 15, 8, 0, 0)));
        clock::reset();
    }

    #[test]
    fn test_is_yesterday() {
        clock::set_test_now(make_utc(2025, 6, 15, 12, 0, 0));
        assert!(is_yesterday(make_utc(2025, 6, 14, 8, 0, 0)));
        assert!(!is_yesterday(make_utc(2025, 6, 15, 8, 0, 0)));
        clock::reset();
    }

    #[test]
    fn test_is_work_day() {
        // 2025-06-16 is a Monday
        assert!(is_work_day(make_utc(2025, 6, 16, 12, 0, 0)));
        // 2025-06-15 is a Sunday
        assert!(!is_work_day(make_utc(2025, 6, 15, 12, 0, 0)));
    }

    #[test]
    fn test_weekday_name() {
        // 2025-06-16 is a Monday
        assert_eq!(weekday_name(make_utc(2025, 6, 16, 12, 0, 0)), "Monday");
        // 2025-06-15 is a Sunday
        assert_eq!(weekday_name(make_utc(2025, 6, 15, 12, 0, 0)), "Sunday");
    }

    #[test]
    fn test_date_str() {
        let dt = make_utc(2025, 6, 15, 12, 0, 0);
        let s = date_str(dt);
        assert!(s.contains("2025"));
    }

    #[test]
    fn test_pretty_date_today() {
        clock::set_test_now(make_utc(2025, 6, 15, 12, 0, 0));
        assert_eq!(pretty_date(make_utc(2025, 6, 15, 8, 0, 0), false), "today");
        clock::reset();
    }

    #[test]
    fn test_pretty_date_tomorrow() {
        clock::set_test_now(make_utc(2025, 6, 15, 12, 0, 0));
        assert_eq!(
            pretty_date(make_utc(2025, 6, 16, 8, 0, 0), false),
            "tomorrow"
        );
        clock::reset();
    }

    #[test]
    fn test_pretty_date_yesterday() {
        clock::set_test_now(make_utc(2025, 6, 15, 12, 0, 0));
        assert_eq!(
            pretty_date(make_utc(2025, 6, 14, 8, 0, 0), false),
            "yesterday"
        );
        clock::reset();
    }

    #[test]
    fn test_is_midnight() {
        let local_midnight = Local
            .from_local_datetime(
                &chrono::NaiveDate::from_ymd_opt(2025, 6, 15)
                    .unwrap()
                    .and_hms_opt(0, 0, 0)
                    .unwrap(),
            )
            .single()
            .unwrap()
            .with_timezone(&Utc);
        assert!(is_midnight(local_midnight));
    }

    #[test]
    fn test_this_weeks_monday() {
        // 2025-06-18 is a Wednesday
        let wed = make_utc(2025, 6, 18, 12, 0, 0);
        let monday = this_weeks_monday(wed);
        let local_monday = monday.with_timezone(&Local);
        assert_eq!(local_monday.weekday(), Weekday::Mon);
        assert_eq!(local_monday.day(), 16);
    }

    #[test]
    fn test_next_monday() {
        // 2025-06-18 is a Wednesday
        let wed = make_utc(2025, 6, 18, 12, 0, 0);
        let nm = next_monday(wed);
        let local_nm = nm.with_timezone(&Local);
        assert_eq!(local_nm.weekday(), Weekday::Mon);
        assert_eq!(local_nm.day(), 23);
    }

    #[test]
    fn test_is_after_today() {
        clock::set_test_now(make_utc(2025, 6, 15, 12, 0, 0));
        assert!(is_after_today(make_utc(2025, 6, 16, 8, 0, 0)));
        assert!(!is_after_today(make_utc(2025, 6, 15, 8, 0, 0)));
        assert!(!is_after_today(make_utc(2025, 6, 14, 8, 0, 0)));
        clock::reset();
    }

    #[test]
    fn test_is_this_week() {
        // 2025-06-18 is a Wednesday, so Monday is 2025-06-16
        clock::set_test_now(make_utc(2025, 6, 18, 12, 0, 0));
        let monday = this_weeks_monday(clock::now());
        let next_mon = next_monday(clock::now());
        let sunday = monday + Duration::days(6);

        assert!(is_this_week(monday));
        assert!(is_this_week(sunday));
        assert!(!is_this_week(next_mon));
        clock::reset();
    }

    #[test]
    fn test_is_next_7_days() {
        clock::set_test_now(make_utc(2025, 6, 15, 12, 0, 0));
        let now = clock::now();
        let yesterday = now - Duration::hours(24);
        let tomorrow = now + Duration::hours(24);
        let in_7_days = now + Duration::hours(24 * 7);

        assert!(is_next_7_days(now));
        assert!(!is_next_7_days(yesterday));
        assert!(is_next_7_days(tomorrow));
        assert!(!is_next_7_days(in_7_days));
        clock::reset();
    }
}
